Refactor HCO controller reconcile function

rwsu opened the pull request:
Broke up the CR creation code in reconcile into functions.
The functions are now called in parallel.

rwsu commented:
I would like some feedback on this refactoring. Whether parallelism is desired. There is still a lot of duplicate code. I have not found a common interface/type between the objects that are returned by the newX functions. Having one would simplify the code quite a bit.

djzager commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -302,6 +225,25 @@ func newKubeVirtConfigForCR(cr *hcov1alpha1.HyperConverged) *corev1.ConfigMap {
 }
 
+func createKubeVirtConfigCRAndSetOwnership(instance *hcov1alpha1.HyperConverged, request reconcile.Request, r *ReconcileHyperConverged, c chan CreateCRResult) {
+	kvConfig := newKubeVirtConfigForCR(instance)
```
is it possible to simply have one function, like `createCrAndSetOwnership` that (in addition to what I see here) takes the name (ie. `KubeVirtConfig`) and the CR generator function (ie. `newKubeVirtConfigForCR`)?

rwsu replied:
Let me look into this again. I had tried before, but didn't find a compatible type. Looking at the add function simplification you submitted, I wonder if we can also use runtime.Object here.

rwsu commented:
Reworked the code to follow some of the patterns that I saw being used in the CDI operator. 
Added tests to verify that the resources being created by HCO are actually present after a reconcile.

rthallisey commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -107,261 +109,91 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
+	var hasError bool
+	// Loop through the results, until an error is found. Once an error is found, don't change result or crErr, but
+	// drain the results channel. If no errors are found, return the last result.
+	for r := range c {
```
Will `c` always have all the results from the channel when this is executed?  I'm not sure it will because `wg.Wait()` is run in a thread.

rwsu replied:
Yes, c will always have all the results. wg.Wait() doesn't run until the WaitGroup is drained. That occurs when the last deferred wg.Done() is called when the last goroutine in the loop above completes.

To be honest, I think making this parallel may be overkill at this point. I don't know if we gain that much or at all by making it parallel. It was difficult to assess the difference in speed when I observed the CRs being created before and after this refactor. This may be premature optimization. 

Would you be ok if I rip the parallelism out and but leave the rest in place?

rthallisey replied:
I'm not as concerned with speed as I am with resiliency.  As long as we don't continuously fail on the first error we see, it's a big upgrade.  Also, 👍  to more tests.

rthallisey replied:
Makes sense to drop the parallelism.  We can come back to it later.

rwsu replied:
Ok removed.

lveyde commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -119,79 +118,58 @@ type ResultFromCRCreate struct {
+				results = append(results, ResultFromCRCreate{result: reconcile.Result{}, err: err})
+			}
+
+			// TODO: common-templates and cdi fails the Get check above and appears to still be missing.
```
I noticed in the past that the code that queries for CDI fails, unless an empty string ("") is used for the namespace. It may be related.

rwsu replied:
I like the flow of check and if not exist, create. It would be good to figure out why the GET requests aren't working correctly. @rthallisey, do you know where they are globally scoped? 

thanks @lveyde, I checked and setting namespace to "" for CDI fixes it. Is empty string ok as the namespace for CDI?

Need to figure out what is going on with common-templates.

rthallisey replied:
Ya an empty string is how you're suppose to reference a globally scoped CRD.  We'll need keep a list of those and have different check for them. 

>@rthallisey, do you know where they are globally scoped?
>
Not sure I understand the question.

The empty string didn't also GET the common-templates?

rwsu replied:
@rthallisey let me rephrase, how do you know if something is globally scoped?

The empty string doesn't work for common-templates.

djzager replied:
If the something you are referring to is a CRD. Then you can look at the `spec.scope` of the CRD.

For example, [cluster network addons](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/master/deploy/converged/crds/cna.crd.yaml#L13) or [kubevirt](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/master/deploy/converged/crds/kubevirt.crd.yaml#L24)

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

djzager commented on pkg/controller/hyperconverged/factory.go:
```diff
@@ -0,0 +1,149 @@
+// The set of resources managed by the HCO
+func (r *ReconcileHyperConverged) getAllResources(cr *hcov1alpha1.HyperConverged, request reconcile.Request) []runtime.Object {
+	return []runtime.Object{
+		newKubeVirtConfigForCR(cr, request.Namespace),
```
do we really need for these to be function calls when we can just list them here?

```
func () getAllResources(...) []runtime.Object {
	labels := map[string]string{
		"app": cr.Name,
	}
	return []runtime.Object{
		&corev1.ConfigMap{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "kubevirt-config",
				Labels:    labels,
				Namespace: namespace,
			},
			Data: map[string]string{
				"feature-gates": "DataVolumes,SRIOV,LiveMigration,CPUManager,CPUNodeDiscovery",
			},
		},
	       ...
	}
}
```

djzager replied:
This contradicts my recommendation below :smile: 

I think having a collection of functions that individually reconcile their respective resources to the state they should be in is the correct way to go.

djzager commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -121,325 +121,72 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
-		log.Info("Existing resource wasn't deployed by HCO, ignoring", "Kind", resource.GetObjectKind())
-		return reconcile.Result{}, nil
-	}
+	resources := r.getAllResources(cr, request)
```
To be honest, I find [this portion of the knicluster_controller reconcile](https://github.com/mhrivnak/kni-operator/blob/master/pkg/controller/knicluster/knicluster_controller.go#L272-L281) to be similar to what you are trying to accomplish here while being somewhat easier to follow and reason about.

Maybe we can update the `newKubevirtTemplateValidatorForCR` type functions to instead ensure that a given resource is as it should be. Then in the reconcile we just loop through those functions.

djzager replied:
Another advantage of this approach is that if it gets more complex than "create if it doesn't exist" then that complexity is in the function where that logic belongs.

rwsu replied:
@djzager I submitted an update. It should look more like the knicluster_controller now.

rthallisey commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -121,325 +121,115 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
-	labels := map[string]string{
-		"app": cr.Name,
+func (r *ReconcileHyperConverged) ensureResourceExists(instance *hcov1alpha1.HyperConverged, logger logr.Logger, request reconcile.Request, desiredRuntimeObj runtime.Object) error {
+	desiredMetaObj := desiredRuntimeObj.(metav1.Object)
```
I'm not familiar with what you're doing in `desiredRuntimeObj.(metav1.Object)`.   Specially, how to the parenthesis around `metav1.Object` effect this call?

rwsu replied:
The underlying object implements multiple interfaces. It comes in as runtime.Object. It is "cast" as metav1.Object here so that the Name and Namespace functions are accessible.

rthallisey commented on pkg/controller/hyperconverged/factory.go:
```diff
@@ -0,0 +1,149 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cdi-" + cr.Name,
+			Labels:    labels,
+			Namespace: namespace,
```
Shouldn't this always be `""`?  Line 154 calls `cdi := newCDIForCR(instance, request.Namespace)`, which may not always be correct.  Maybe we should create some `const`s with NS values and spread them around?

rwsu replied:
Good catch. Yes, constants are a good idea. Added.

rwsu commented on pkg/controller/hyperconverged/hyperconverged_controller.go:
```diff
@@ -121,325 +121,115 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
+			return err
+		}
+
+		// TODO: common-templates fails the Get check above and appears to still be missing.
```
Appears so.

The CR is in the openshift ns:

````
[rwsu@localhost converged]$ kubectl get kubevirtcommontemplatesbundle -n openshift
NAME                                      AGE
common-templates-hyperconverged-cluster   4m51s
````

request.Namespace is kubevirt-hyperconverged, but key.namespace is openshift. This may be why the Get is not able to find it.

What is interesting is KWebUI which doesn't specify a namespace in the CR can be queried/Get using any namespace.

rthallisey replied:
`Kwebui`, `CDI`, and `NetworkAddons` are all cluster scoped, so that makes sense.
```
...
  names:
    kind: KWebUI
    listKind: KWebUIList
    plural: kwebuis
    singular: kwebui
  scope: Cluster
  version: v1alpha1
```

Do you think this `TODO` is no longer needed?

rthallisey commented:
ci test please

rthallisey commented:
ci add to whitelist

rthallisey commented:
@rwsu I think this is ready to go.  Just one last nit

rthallisey commented:
nice job @rwsu
